### PR TITLE
MacOS Build Asset and Workflow Optimization

### PR DESCRIPTION
**Pull request by vgrassia**
This PR updates the MacOS workflow with caching and separate jobs for asset generation.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -312,78 +345,412 @@ jobs:
...
         shell: pwsh
         run: ./scripts/safari-build.ps1 -skipcheckout -skipoutcopy
 
-      - name: Load Safari extension for .dmg
-        if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/rc'
+  macos-package-github:
+    name: MacOS Package GitHub Release Assets
+    runs-on: macos-latest
+    needs: macos-build
+    if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/rc' || github.ref == 'refs/heads/better-build-assets'
```
Comment (joseph-flinn) [inline .github/workflows/build.yml:375]: could you remove the testing code?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -312,78 +345,412 @@ jobs:
...
         with:
           name: Bitwarden-${{ env.PACKAGE_VERSION }}.dmg
           path: ./dist/Bitwarden-${{ env.PACKAGE_VERSION }}.dmg
           if-no-files-found: error
 
+  macos-package-mas:
+    name: MacOS Package Prod Release Asset
+    runs-on: macos-latest
+    needs: macos-build
+    if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/rc' || github.ref == 'refs/heads/better-build-assets'
```
Comment (joseph-flinn) [inline .github/workflows/build.yml:506]: could you remove the testing code?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -312,78 +345,412 @@ jobs:
...
+          APPLE_ID_PASSWORD: ${{ secrets.APPLE_ID_PASSWORD }}
+
+      - name: Set up provisioning profiles
+        run: ./.github/scripts/macos/setup-profiles.ps1
+        shell: pwsh
+
+      - name: Increment version
+        run: ./.github/scripts/macos/increment-version.ps1
+        shell: pwsh
+
```
Comment (joseph-flinn) [inline .github/workflows/build.yml:446]: These are essentially setting environment variables to use later. Thoughts on doing this in the build step and setting them as outputs and then grabbing them in these jobs?

We can always do this when we come back around to do more cleanup

**Review by joseph-flinn — APPROVED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (joseph-flinn) [inline .github/workflows/build.yml:375]: There's some testing code in here

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -312,78 +345,412 @@ jobs:
...
+          MACDEV_CERT_PASSWORD: ${{ secrets.MACDEV_CERT_PASSWORD }}
+          APPLE_ID_PASSWORD: ${{ secrets.APPLE_ID_PASSWORD }}
+
+      - name: Set up provisioning profiles
+        run: ./.github/scripts/macos/setup-profiles.ps1
+        shell: pwsh
+
+      - name: Increment version
+        run: ./.github/scripts/macos/increment-version.ps1
+        shell: pwsh
```
Comment (joseph-flinn) [inline .github/workflows/build.yml:445]: These are essentially setting environment variables to use later. Thoughts on doing this in the build step and setting them as outputs and then grabbing them in these jobs?

We can always do this when we come back around to do more cleanup

File: .github/workflows/build.yml (hunk 2 above)
Comment (joseph-flinn) [inline .github/workflows/build.yml:506]: There's some testing code in here

**Review by joseph-flinn — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 4 above)
Comment (vgrassia) [reply .github/workflows/build.yml:445]: I think some of them copy files around as well.  I'd vote to leave it for later cleanup.

**Review by joseph-flinn — APPROVED** [review]

**Pull request merged**